python/pdo/service_client/servicedatabase.py

```diff
@@ -0,0 +1,292 @@
+            raise Exception('Cannot create service client : Invalid Database, id found multiple times in the database')
+        elif entries[0]['url'] is None:
+            logger.exception('Unable to find service url for given id. Cannot create serivce client without the url')
+            raise Exception('Unable to find service url for given id. Cannot create serivce client without the url')
+
+        
+        if self._service_type=='eservice':
+            client =  self.get_eservice_client(entries[0]['url'])
+            service_info = self.get_info(url = entries[0]['url'])
+            if service_info['id'] != id:
+                logger.exception('Enclave hosted by the eservice does not match the one found in database: Cannot create service client')
+                raise Exception('Enclave hosted by the eservice does not match the one found in database: Cannot create service client')
+        else:
+            logger.exception('Unsupported operation. Service client by id is supported only for eservice and not for sservice')
+            raise Exception('Unsupported operation. Service client by id is supported only for eservice and not for sservice')
+
+        return client
+    
+    #--------------------------------------------
+    #--------------------------------------------
+    
+    def add_new_info(self, name, id = None, url = None):
+        """ Add a new entry to the data base. If any one field of the new entry is already in the database, an exception will be raised."""
+
+        if name is None:
+            logger.exception('Cannot add entry to service database: Need a name for the entry')
+            raise Exception('Cannot add entry to service database: Need a name for the entry')
+        elif name in self.data.keys():
+            logger.exception('Cannot add entry to service database: Name already present, give a new name for the entry')
+            raise Exception('Cannot add entry to service database: Name already present, give a new name for the entry')
+        
+        if url is not None:
+            entries = list(filter(lambda entry: are_the_urls_same(entry['url'], url) , self.data.values()))
+            if len(entries) >0:
+                logger.exception('Cannot add entry to service database : url already present in the database')
+                raise Exception('Cannot add entry to service database : url already present in the database')
+
+        if id is not None:
+            entries = list(filter(lambda entry: entry['id']==id , self.data.values()))
+            if len(entries) >0:
+                logger.exception('Cannot add entry to service database : id already present in the database')
+                raise Exception('Cannot add entry to service database : id already present in the database')
+
+        # all good to add the new entry
+        self.data[name] = {'id' : id, 'url': url}
+
```

Comment:
you have a LOT of complexity here that seems unnecessary.
when you load the database, create two dictionaries that serve as your indexes into the data. for example have something that maps a name to a URL, and one that maps a URL to an enclave ID. then you can have three simple look ups rather than going through these fairly convoluted checks.

Reply:
agreed that having secondary dictonaries is a better solution if there are mutliple db operations of a large db from the same client during the session.